### PR TITLE
bugfix/16782-setdata-unsorted-joinby

### LINKED ISSUE #16782: Highmaps - `joinBy` is not respected in series update
#### Expected behaviour
Workaround with an empty point on init: https://jsfiddle.net/BlackLabel/j3vb21qm/

#### Actual behaviour
Points are not matched by `hc-key`.

#### Live demo with steps to reproduce
1. Open jsfiddle: https://jsfiddle.net/BlackLabel/j2qrk1pL/
2. Wait one second
3. Check the most right-bottom point - it is Hessen and should be Bayern

#### Product version
Highmaps 9.3

#### Affected browser(s)
All

_Internal note:_
Regression bug since Highmaps 9.3

### PR DESCRIPTION
**Pull request by raf18seb**
Fixed #16782, point's `joinBy` was wrong when updating data with an unsorted order.

### CODE REVIEW

**Review by DJTechnoo — APPROVED** [review]

**Comment by TorsteinHonsi** [comment]
Thanks! Can you explain why sorting matters? I would expect the `joinBy` logic to make sure each new `data` option to be linked to existing points (from `mapData`) regardless of its position in the array?

**Comment by raf18seb** [comment]
Sure!

If `series.data` is an empty array initially, then map points are created from the map source in a given order.
Then, when setting a real data with a different order than source map data but **with the same data.length**, it reaches this part of code: https://github.com/highcharts/highcharts/blob/eb5a3f607eafd088fd033cee4af2a6b8b48adf0e/ts/Core/Series/Series.ts#L1216-L1223

Here the order in `oldData` and `data` is different and points are mixed up.

The first solution that came to my mind was to set `updatePoints` flag to `false` in `setData(data, ...)`, but it doesn't solve a case with `chart.update({ series: [{ data: newData }] });`

Then I was thinking about adding new condition here to this `if` statement: https://github.com/highcharts/highcharts/blob/eb5a3f607eafd088fd033cee4af2a6b8b48adf0e/ts/Core/Series/Series.ts#L1342-L1351
but `dataLength !== oldDataLength` doesn't make sense, and it touches Highcharts code not related to Maps.

So I decided to simply sort the data by `joinBy` manually.

I'm open to other solution if there is a better one

**Comment by TorsteinHonsi** [comment]
Thanks for the explanation! The order or keys I think is a fragile metric - as you can see it falls apart if we add one mismatch: https://jsfiddle.net/highcharts/qLsb9n5t/

I'd rather go with your second proposal, to extend the condition for when to run `this.updateData`. If we add `!series.processedData` it works. As you say it touches Highcharts code not related to Maps. And as you know I am probably the cheapest person on the planet when it comes to bloating the core, but this is only a few bytes. We do the same for `!series.hasGroupedData` and `!series.isSeriesBoosting`. 

In fact we could probably avoid touching the core by adding something like this to the MapSeries (and map bubble?):

```ts
public updateData(data) {
    if (this.processedData) {
         return false;
    }
    super();
}
```

**Comment by raf18seb** [comment]
Nice, thanks!

It's working fine, but it revealed another problem with setData for mapbubble with `updatePoints: false`.
Reported as a separate bug here: https://github.com/highcharts/highcharts/issues/17112

**Review by KacperMadej — APPROVED** [review]

File: ts/Core/Chart/Chart.ts [hunk 1]
```diff
@@ -3383,9 +3382,7 @@ class Chart {
         if (updateAllSeries) {
             chart.getSeriesOrderByLinks().forEach(function (series): void {
                 // Avoid removed navigator series
-                if (series.chart) {
-                    series.update({}, false);
-                }
+                series.update({}, false);
```
Comment (TorsteinHonsi) [inline ts/Core/Chart/Chart.ts:3385]: Why was the check for `series.chart` removed? What consequences may it have? The comment (which is now dangling) describes a removed navigator series.

**Review by TorsteinHonsi — COMMENTED** [review]
Check out the removed guard.

File: ts/Core/Chart/Chart.ts (hunk 1 above)
Comment (raf18seb) [reply ts/Core/Chart/Chart.ts]: Sorry, I rushed with it. I saw an unusual `if` check, I thought it was unnecessary and removed it. Don't know the consequences :(
Already restored.

**Review by TorsteinHonsi — APPROVED** [review]

**Pull request merged**